core/src/main/java/org/apache/iceberg/util/PartitionMap.java

```diff
@@ -0,0 +1,265 @@
+        return removeKey((Integer) first, (StructLike) second);
+      }
+    }
+
+    return null;
+  }
+
+  public V removeKey(int specId, StructLike struct) {
+    Map<StructLike, V> partitionMap = partitionMaps.get(specId);
+    return partitionMap != null ? partitionMap.remove(struct) : null;
+  }
+
+  @Override
+  public void clear() {
+    partitionMaps.clear();
+  }
+
+  @Override
+  public Set<Pair<Integer, StructLike>> keySet() {
+    PartitionSet keySet = PartitionSet.create(specs);
+
+    for (Entry<Integer, Map<StructLike, V>> specIdAndPartitionMap : partitionMaps.entrySet()) {
+      int specId = specIdAndPartitionMap.getKey();
+      Map<StructLike, V> partitionMap = specIdAndPartitionMap.getValue();
+      for (StructLike partition : partitionMap.keySet()) {
+        keySet.add(specId, partition);
+      }
+    }
+
+    return Collections.unmodifiableSet(keySet);
+  }
+
+  @Override
+  public Collection<V> values() {
+    List<V> values = Lists.newArrayList();
+
+    for (Map<StructLike, V> partitionMap : partitionMaps.values()) {
+      values.addAll(partitionMap.values());
+    }
+
+    return Collections.unmodifiableCollection(values);
+  }
+
+  @Override
+  public Set<Entry<Pair<Integer, StructLike>, V>> entrySet() {
+    Set<Entry<Pair<Integer, StructLike>, V>> entrySet = Sets.newHashSet();
+
+    for (Entry<Integer, Map<StructLike, V>> specIdAndPartitionMap : partitionMaps.entrySet()) {
+      int specId = specIdAndPartitionMap.getKey();
+      Map<StructLike, V> partitionMap = specIdAndPartitionMap.getValue();
+      for (Entry<StructLike, V> structAndValue : partitionMap.entrySet()) {
+        StructLike struct = structAndValue.getKey();
+        V value = structAndValue.getValue();
+        entrySet.add(new PartitionEntry<>(specId, struct, value));
+      }
+    }
+
+    return Collections.unmodifiableSet(entrySet);
+  }
+
+  public V computeIfAbsent(int specId, StructLike struct, Supplier<V> valueSupplier) {
+    Map<StructLike, V> partitionMap = partitionMaps.computeIfAbsent(specId, this::newPartitionMap);
+    return partitionMap.computeIfAbsent(struct, key -> valueSupplier.get());
+  }
+
+  private Map<StructLike, V> newPartitionMap(int specId) {
+    PartitionSpec spec = specs.get(specId);
+    return StructLikeMap.create(spec.partitionType());
+  }
+
+  @Override
+  public String toString() {
+    return partitionMaps.entrySet().stream()
+        .flatMap(this::toStringStream)
+        .collect(Collectors.joining(", ", "[", "]"));
+  }
+
+  private Stream<String> toStringStream(Entry<Integer, Map<StructLike, V>> entry) {
+    PartitionSpec spec = specs.get(entry.getKey());
+    return entry.getValue().entrySet().stream().map(innerEntry -> toString(spec, innerEntry));
+  }
+
+  private String toString(PartitionSpec spec, Entry<StructLike, V> entry) {
+    StructLike struct = entry.getKey();
+    V value = entry.getValue();
+    return spec.partitionToPath(struct) + " -> " + (value == this ? "(this Map)" : value);
```

Comment:
Not sure I follow this string creation

/foo=1/ -> (this map)

Are we assuming we would get nested PartitionMaps?

Reply:
I would rather we just ban putting the map into itself?

Reply:
Banning would mean adding extra logic on each put, which I'd probably avoid (to be honest, it is highly unlikely to hit this in practice).

I copied this approach from Java `AbstractMap` that does a similar trick.

```
public String toString() {
    Iterator<Entry<K,V>> i = entrySet().iterator();
    if (! i.hasNext())
        return "{}";

    StringBuilder sb = new StringBuilder();
    sb.append('{');
    for (;;) {
        Entry<K,V> e = i.next();
        K key = e.getKey();
        V value = e.getValue();
        sb.append(key   == this ? "(this Map)" : key);
        sb.append('=');
        sb.append(value == this ? "(this Map)" : value);
        if (! i.hasNext())
            return sb.append('}').toString();
        sb.append(',').append(' ');
    }
}
```

Reply:
We can keep with tradition